crates/ra_hir/src/ty.rs

```diff
@@ -1042,12 +1046,26 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             },
             Expr::Tuple { exprs } => {
                 let mut ty_vec = Vec::with_capacity(exprs.len());
-                for arg in exprs.iter() {
-                    ty_vec.push(self.infer_expr(*arg, &Expectation::none())?);
+                for expr in exprs.iter() {
+                    ty_vec.push(self.infer_expr(*expr, &Expectation::none())?);
                 }
 
                 Ty::Tuple(Arc::from(ty_vec))
             }
+            Expr::Array { exprs } => {
+                let mut ty = self.insert_type_vars(expected.ty.clone());
+
+                for (i, expr) in exprs.iter().enumerate() {
+                    if i == 0 {
```

Comment:
oops, this is not good implementation.
If getting `[1, y, 2, 3]`, it must inferes `[_]`.
I'll fix it.

Reply:
You don't need to use the inferred type from the subexpressions. If you just pass `Expectation::has_type(ty)`, all the element expressions will be unified to the same type. No need for special casing the first element either :)